Digital Credentials: conditionally enable the API based on entitlement

marcoscaceres opened the pull request:
#### bad163661a0fd171bcf995e1baa3e67709723581
<pre>
Digital Credentials: conditionally enable the API based on entitlement
<a href="https://rdar.apple.com/152637111">rdar://152637111</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=294889">https://bugs.webkit.org/show_bug.cgi?id=294889</a>

Reviewed by Abrar Rahman Protyasha and Pascoe.

Conditionally enable the Digital Credentials API based on the appropriate entitlements.

* Source/WebKit/Shared/WebPreferencesDefaultValues.cpp:
(WebKit::defaultDigitalCredentialsEnabled):

Canonical link: <a href="https://commits.webkit.org/296643@main">https://commits.webkit.org/296643@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5c50a53f7cbfe91fcfcdfb4220fc72e503affbb2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/109068 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/28729 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/19153 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/114279 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/59372 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/29411 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/37294 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/82885 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/112016 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/23392 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/98234 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/63330 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22793 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58965 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/92762 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/16419 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/117397 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/36117 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/26693 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/91899 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/36487 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/94498 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/91705 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23369 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/36614 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/14368 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/31968 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/36014 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/41519 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/35707 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/39046 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/37393 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebKit/Shared/WebPreferencesDefaultValues.cpp:
```diff
@@ -245,7 +245,9 @@ bool defaultGamepadVibrationActuatorEnabled()
-    return true;
+    Ref webProcess = WebProcess::singleton();
+    bool enabled = webProcess->parentProcessHasEntitlement("com.apple.developer.web-browser"_s) || webProcess->parentProcessHasEntitlement("com.apple.developer.identity-document-services.web-presentment-controller"_s);
+    return enabled;
```
Looking at the other code in this file I suspect you want and maybe you should also use `dispatch_once`?
```suggestion
    static bool hasEntitlement = webProcess->parentProcessHasEntitlement("com.apple.developer.web-browser"_s) || webProcess->parentProcessHasEntitlement("com.apple.developer.identity-document-services.web-presentment-controller"_s);
    return hasEntitlement;
```

marcoscaceres replied:
Moved to using dispatch_once

annevk reviewed (commented):
Also, is this never called outside the web process?

pascoej commented on Source/WebKit/Shared/WebPreferencesDefaultValues.cpp:
```diff
@@ -37,7 +37,7 @@
 #endif
 #endif
 
-#if ENABLE(MEDIA_SESSION_COORDINATOR)
```
could this be `|| HAVE(DIGITAL_CREDENTIALS_UI)`?

aprotyas replied:
Same comment

marcoscaceres replied:
It could. I'll switch it.

aprotyas commented on Source/WebKit/Shared/WebPreferencesDefaultValues.cpp:
```diff
@@ -245,7 +245,15 @@ bool defaultGamepadVibrationActuatorEnabled()
+        if (isInWebProcess())
+            enabled = WebProcess::singleton().parentProcessHasEntitlement("com.apple.developer.web-browser"_s) || WebProcess::singleton().parentProcessHasEntitlement("com.apple.developer.identity-document-services.web-presentment-controller"_s);
+        else
+            enabled = WTF::processHasEntitlement("com.apple.developer.web-browser"_s) || WTF::processHasEntitlement("com.apple.developer.identity-document-services.web-presentment-controller"_s);
```
Optional suggestion: this helps ever so slightly with avoiding repetition of the long entitlements.

```cpp
auto entitlementChecker = [inWebProcess = isInWebProcess()](auto entitlement) {
    if (inWebProcess)
        return WebProcess::singleton().parentProcessHasEntitlement(entitlement);
    return WTF::processHasEntitlement(entitlement);
};

enabled = entitlementChecker("com.apple.developer.web-browser"_s) || entitlementChecker("com.apple.developer.identity-document-services.web-presentment-controller"_s);
```

marcoscaceres replied:
Like it. Will add.

marcoscaceres committed:
Digital Credentials: conditionally enable the API based on entitlement
rdar://152637111
https://bugs.webkit.org/show_bug.cgi?id=294889

Reviewed by Abrar Rahman Protyasha and Pascoe.

Conditionally enable the Digital Credentials API based on the appropriate entitlements.

* Source/WebKit/Shared/WebPreferencesDefaultValues.cpp:
(WebKit::defaultDigitalCredentialsEnabled):

Canonical link: https://commits.webkit.org/296643@main
